text/0000-decorators.md

```diff
@@ -0,0 +1,1212 @@
+  })
+})
+
+class Person {
+  cache = {};
+}
+```
+
+#### Preventing Incoherent Usage
+
+Making the `ComputedPropertyConfig` optional opens up lots of room for
+accidents. A computed property without a getter or setter does not make sense,
+nor does a computed propery with _two_ getters or setters. The new decorator
+will assert at _decorator application_ time to ensure it is being used
+correctly:
+
+```js
+import EmberObject, { computed } from '@ember/object';
+
+// This will throw because the user attempted to define a CP without a getter
+const Person = EmberObject.extend({
+  fullName: computed()
+});
+
+// This will also throw because it is missing a getter
+class Person {
+  @computed('firstName', 'lastName')
+  fullName;
+}
+
+// This will throw because a getter was already defined
+class Person {
+  @computed('firstName', 'lastName', {
+    get() {
+      return `${this.firstName} ${this.lastName}`;
+    }
+  })
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+Effectively, if `computed` is passed a `ComputedPropertyConfig`, it returns a
+class field decorator. Otherwise, it returns an accessor (getter/setter)
+decorator.
+
+#### Property Modifiers
+
+Almost all computed property modifiers have been deprecated at this point, but
+they are still in use today and will still be available until Ember v4. As such,
+their syntax needs to remain available and unchanged:
+
+```js
+import EmberObject, { computed } from '@ember/object';
+
+const Person = EmberObject.extend({
+  fullName: computed('firstName', {
+    get() {
+      return `${this.firstName} ${this.lastName}`;
+    }
+  }).readOnly().volatile().property('lastName') // a very strange combination
+});
+```
+
+The decorator returned from `computed()` will need to have these chainable
+methods available, and they will need to set the state of the decorator. This
+should not be too difficult to accomplish.
+
+Usage in native decorator syntax is a little bit trickier. In the current
+proposal, only simple chaining is allowed in a decorator invocation. You may not
+chain on the result of a function:
+
+```js
+import { computed } from '@ember/object';
+
+const fullName = computed('firstName', 'lastName', function() {
+  return `${this.firstName} ${this.lastName}`;
+});
+
+class Person {
+  @computed('firstName', 'lastName').readOnly() // this is invalid
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+
+  @fullName.readOnly() // this is valid because it's a simple chain
+  otherFullName;
+}
+```
+
+Luckily, there is one other form of invocation which is available - wrapping the
+entire decorator expression in parenthesis:
+
+```js
+import { computed } from '@ember/object';
+
+class Person {
+  @(computed('firstName', 'lastName').readOnly()) // this is valid
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+This is clearly _not_ ideal for commonly used features, and while `volatile()`
```

Comment:
We can tackle this in a follow-on discussion or the official RFC when you open it, but it strikes me that this might be a nice thing to build as an optional feature (enabled by default in v3 blueprints, including Octane), so that apps and addons which wish to opt out of it (and the associated increase in size) can do so. For example, in our application, we *never* need to be using these macros—there is exactly one place where we have one, and we've wanted to remove it for a bit and just haven't gotten around to it yet.